luarules/gadgets/unit_weapon_smart_select_helper.lua

```diff
@@ -0,0 +1,216 @@
+local mathMax = math.max
+
+--tables
+local smartUnits = {}
+local unitDefData = {}
+
+function gadget:Initialize()
+	local units = Spring.GetAllUnits()
+	local spGetUnitDefID = Spring.GetUnitDefID
+	for i = 1, #units do
+		gadget:UnitCreated(units[i], spGetUnitDefID(units[i]))
+	end
+end
+
+for unitDefID, def in ipairs(UnitDefs) do
+	if def.weapons then
+		local weapons = def.weapons
+		for weaponNumber, weaponData in pairs(weapons) do
+			local weaponDefID = weapons[weaponNumber].weaponDef
+			if WeaponDefs[weaponDefID] and WeaponDefs[weaponDefID].customParams then
+				if WeaponDefs[weaponDefID].customParams.smart_preferred_weapon then
+					unitDefData[unitDefID] = unitDefData[unitDefID] or {}
+					unitDefData[unitDefID].preferredWeapon = weaponNumber
+					unitDefData[unitDefID].failedToFireFrameThreshold = WeaponDefs[weaponDefID].customParams
+						.smart_error_frames or
+						mathMax(WeaponDefs[weaponDefID].reload * failedToFireMultiplier, minimumFailedToFireFrames)
+					if def.speed and def.speed ~= 0 then
+						unitDefData[unitDefID].canMove = true
+					end
+				end
+				if WeaponDefs[weaponDefID].customParams.smart_deferred_weapon then
+					unitDefData[unitDefID] = unitDefData[unitDefID] or {}
+					unitDefData[unitDefID].deferredWeapon = weaponNumber
+				end
+				if WeaponDefs[weaponDefID].customParams.smart_trajectory_checker then
+					unitDefData[unitDefID] = unitDefData[unitDefID] or {}
+					unitDefData[unitDefID].trajectoryCheckWeapon = weaponNumber
+				end
+			end
+		end
+	end
+end
+
+
+function gadget:UnitCreated(unitID, unitDefID)
+	if unitDefData[unitDefID] then
+		smartUnits[unitID] = {
+			unitDefID = unitDefID,
+			setStateScriptID = Spring.GetCOBScriptID(unitID, "SetAimingState"),
+			aggroBias = 0,
+			failedShotFrame = 0,
+			errorTallyMultiplier = 0,
+		}
+		spCallCOBScript(unitID, smartUnits[unitID].setStateScriptID, 0, unitDefData[unitDefID].preferredWeapon)
+	end
+end
+
+local function failureToFireCheck(attackerID, attackerData, defData)
+	if not attackerData.suspendErrorUntilFrame then return false end
+
+	if attackerData.failedShotFrame < gameFrame - defData.failedToFireFrameThreshold then
+		attackerData.failedShotFrame = mathMax(
+			spGetUnitWeaponState(attackerID, defData.preferredWeapon, 'reloadFrame'),
+			spGetUnitWeaponState(attackerID, defData.deferredWeapon, 'reloadFrame')
+		)
+	end
+
+	if attackerData.failedShotFrame < gameFrame - defData.failedToFireFrameThreshold and
+		gameFrame > attackerData.suspendErrorUntilFrame then
+		return true
+	else
+		return false
+	end
+end
+
+
+local function updateAimingState(attackerID)
+	local attackerData = smartUnits[attackerID]
+	local defData = unitDefData[attackerData.unitDefID]
+
+	local pTargetType, pIsUserTarget, pTarget = spGetUnitWeaponTarget(attackerID, defData.preferredWeapon)
+	local dIsUserTarget, dTarget = select(2, spGetUnitWeaponTarget(attackerID, defData.deferredWeapon))
+
+	local preferredCanShoot = false
+	if pTargetType == 1 then
```

Comment:
What's the magic 1 and 2? Could use named constants.

Reply:
donezo